### PR TITLE
3 increase robustness of the databroker

### LINKED ISSUE #3: Increase robustness of the databroker
We now had the case of a failing agent due to too many callbacks in the databroker. 
Features I am missing:

- debug the current queue size, if it is above a certain threshold or keeps getting bigger
- maybe set a max qsize (would require `Queue` instead of `SimpleQueue`
- Execute the callbacks in threads

I will create a branch and add minimal examples which show the limitations of the databroker. 
I think we should also check how other brokers (e.g. mqtt) handle these cases.

### PR DESCRIPTION
**Pull request by FWuellhorst**
Closes #3

### CODE REVIEW

File: agentlib/core/agent.py [hunk 1]
```diff
@@ -146,9 +144,6 @@ def config(self, config: Union[AgentConfig, FilePath, str, dict]):
 
         self._config = load_config(config, config_type=AgentConfig)
 
-        # Update modules
-        self._register_modules()
-
```
Comment (FWuellhorst) [inline agentlib/core/agent.py:149]: Why did you delete this?

File: agentlib/core/data_broker.py [hunk 2]
```diff
@@ -170,7 +179,7 @@ def __init__(self, logger: CustomLogger):
         self.logger = logger
         self._mapped_callbacks: Dict[Tuple[str, Source], List[BrokerCallback]] = {}
         self._unmapped_callbacks: List[BrokerCallback] = []
-        self._variable_queue = queue.SimpleQueue()
+        self._variable_queue = queue.Queue(maxsize=1000)
```
Comment (FWuellhorst) [inline agentlib/core/data_broker.py:182]: I would make 1000 a parameter of the agent. If you agree, I can implement this.

File: agentlib/core/data_broker.py [hunk 3]
```diff
@@ -203,7 +212,7 @@ def _execute_callbacks(self):
         """
         variable = self._variable_queue.get(block=True)
         qsize = self._variable_queue.qsize()
-        self.logger.debug("Queue fullness: %s", qsize)
+        self.logger.debug("Distribution queue fullness: %s", qsize)
```
Comment (FWuellhorst) [inline agentlib/core/data_broker.py:215]: Only log if percentage of max size is violated?

File: agentlib/core/data_broker.py [hunk 4]
```diff
@@ -393,5 +413,61 @@ def _start_executing_callbacks(self, env: Environment):
...
-        while True:
-            self._execute_callbacks()
+        try:
+            while True:
+                if not self._stop_queue.empty():
+                    err, module_id = self._stop_queue.get()
+                    raise RuntimeError(f"A callback failed in the module {module_id}.") from err
+                self._execute_callbacks()
+        except Exception:
+            raise
```
Comment (FWuellhorst) [inline agentlib/core/data_broker.py:423]: What is the intention of this?

File: agentlib/core/data_broker.py [hunk 5]
```diff
@@ -393,5 +413,61 @@ def _start_executing_callbacks(self, env: Environment):
...
+            _unsafe_no_copy=_unsafe_no_copy,
+            **kwargs,
+        )
+        if callback.module_id not in self._module_queues:
+            self._start_module_thread(callback.module_id)
+        return callback
+
+    def _start_module_thread(self, module_id: str):
+        """Starts a consumer thread for callbacks registered from a module."""
+        module_queue = queue.Queue(maxsize=100)
```
Comment (FWuellhorst) [inline agentlib/core/data_broker.py:447]: Again, this should be set-able

File: agentlib/core/data_broker.py [hunk 6]
```diff
@@ -393,5 +413,61 @@ def _start_executing_callbacks(self, env: Environment):
...
+            kwargs={"queue": module_queue, "module_id": module_id},
+        ).start()
+        self._module_queues[module_id] = module_queue
+
+    def _execute_callbacks_of_module(self, queue: queue.SimpleQueue, module_id: str):
+        """Executes the callbacks associated with a specific module."""
+        try:
+            while True:
+                cb, variable = queue.get(block=True)
+                cb: BrokerCallback
```
Comment (FWuellhorst) [inline agentlib/core/data_broker.py:461]: Really necessary?

File: agentlib/core/data_broker.py [hunk 7]
```diff
@@ -393,5 +413,61 @@ def _start_executing_callbacks(self, env: Environment):
...
+                cb.callback(variable=variable, **cb.kwargs)
+        except Exception as e:
+            self._stop_queue.put((e, module_id))
+            raise e
+
+    def _run_callbacks(self, callbacks: List[BrokerCallback], variable: AgentVariable):
+        """Distributes callbacks to the threads running for each module."""
+        for cb in callbacks:
+            self._module_queues[cb.module_id].put_nowait((cb, variable))
+            self.logger.debug("Queue %s fullness: %s", cb.module_id, self._module_queues[cb.module_id].qsize())
```
Comment (FWuellhorst) [inline agentlib/core/data_broker.py:472]: Same here

File: examples/databroker_limits.py [hunk 8]
```diff
@@ -98,11 +98,11 @@ def faulty_module():
...
+    MAS.run(until=10)
 
 
 if __name__ == '__main__':
     logging.basicConfig(level="DEBUG")
     # slow_module()
-    exploding_modules()
-    # faulty_module()
\ No newline at end of file
+    # exploding_modules()
```
Comment (FWuellhorst) [inline examples/databroker_limits.py]: add to tests

**Review by EserSteffen — APPROVED** [review]
Looks good now.
